Sources/kai/Parser.swift

```diff
@@ -342,6 +355,87 @@ extension Parser {
+            case .lbrace?:
+                cond = nil
+                
+            default:
+                // Make sure an identifier doesn't get turned into a
+                // compound literal
+                let prevState = state
+                defer { state = prevState }
+                state.insert(.disallowCompoundLiteral)
+                
+                cond = try expression()
+            }
+            
+            try consume(.lbrace)
+            
+            var cases: [AstNode] = []
+            var defaultBody: AstNode? = nil
+            
+            while try lexer.peek()?.kind != .rbrace {
+                try consumeTerminators(justNewlines: true)
+                
+                switch try lexer.peek()?.kind {
+                case .keyword(.case)?:
+                    let (_, startLocation) = try consume()
+                    // Equal binding power with `:`
+                    let cond = try expression(10)
+                    
+                    try consume(.colon)
+                    
+                    let terminators: [Lexer.Token] = [.rbrace, .keyword(.case), .keyword(.default)]
+                    let body = try expressions(until: terminators)
+                    let loc = startLocation ..< lexer.location
+                    
+                    cases.append(
+                        .stmtCase(
+                            cond: cond,
+                            body: AstNode.list(body, loc),
+                            loc
+                        )
+                    )
+                    
+                case .keyword(.default)?:
+                    let (_, startLocation) = try consume()
+                    try consume(.colon)
+                    
+                    let terminators: [Lexer.Token] = [.rbrace, .keyword(.case), .keyword(.default)]
+                    let body = try expressions(until: terminators)
+                    let loc = startLocation ..< lexer.location
+                    
+                    guard defaultBody == nil else {
+                        reportError("Duplicate default cases in switch statement", at: startLocation ..< lexer.location)
+                        return AstNode.invalid(startLocation ..< lexer.location)
+                    }
+                    
+                    defaultBody = AstNode.list(body, startLocation ..< lexer.location)
+                    
+                case .rbrace?:
+                    continue
+                    
+                default:
+                    reportError("Unexpected token \(try lexer.peek()!)", at: lexer.location)
+                }
+            }
+            
+            try consume(.rbrace)
+            
+            return AstNode.stmtSwitch(
+                cond: cond,
+                body: .list(cases, location ..< location),
+                defaultBody: defaultBody, location ..< location
+            )
+          
+        case .keyword(.case):
```

Comment:
If we use the subparser stuff I mentioned above, this will become the `case` parser and should check the parser state to ensure a case statement is valid.